src/cpp/src/visual_language/cdpruner/cdpruner.cpp

```diff
@@ -0,0 +1,401 @@
+ * config.relevance_weight = 1.0;
+ *
+ * CDPruner pruner(config);
+ * auto pruned_features = pruner.apply_pruning(visual_features, text_features);
+ * ```
+ */
+
+#include "cdpruner.hpp"
+
+#include <algorithm>
+#include <chrono>
+#include <cstring>
+#include <future>
+#include <iomanip>
+#include <iostream>
+#include <stdexcept>
+#include <thread>
+
+#include "logger.hpp"
+#include "openvino/openvino.hpp"
+#include "utils.hpp"
+
+namespace ov::genai::cdpruner {
+
+CDPruner::CDPruner(const Config& config)
+    : m_config(config),
+      m_relevance_calc(config),
+      m_kernel_builder(config),
+      m_dpp_selector(config) {
+    m_config.update_from_env();
+    validate_config(config);
+}
+
+std::vector<std::vector<size_t>> CDPruner::select_tokens(const ov::Tensor& visual_features,
+                                                         const ov::Tensor& text_features,
+                                                         bool silent) {
+    validate_input_tensors(visual_features, text_features);
+
+    const auto& visual_shape = visual_features.get_shape();
+    size_t batch_size = visual_shape[0];
+    size_t total_tokens = visual_shape[1];
+    size_t feature_dim = visual_shape[2];
+
+    size_t raw_tokens_to_keep = static_cast<size_t>(std::round(total_tokens * (1.0 - m_config.pruning_ratio / 100.0)));
+    // Round up to the next even number of tokens to keep
+    // This is required for DPP OpenCL implementation
+    size_t num_tokens_to_keep = (raw_tokens_to_keep % 2 == 0) ? raw_tokens_to_keep : raw_tokens_to_keep + 1;
+
+    try {
+        std::vector<std::vector<size_t>> selected_tokens;
+
+        // Determine whether to use token splitting strategy
+        // Splitting is introduced to handle large token sequences more efficiently:
+        // 1. Computational efficiency: Greedy DPP algorithm complexity is O(N×M²) where N is total
+        //    tokens and M is tokens to select. Splitting into halves could provide ~4× speedup for large sequences.
+        // 2. Parallel processing: Two halves can be processed independently, enabling parallelization.
+        bool use_splitting = m_config.split_threshold > 0 && total_tokens > m_config.split_threshold;
+
+        size_t split_point = 0;
+        size_t first_half_size = 0;
+        size_t second_half_size = 0;
+
+        if (use_splitting) {
+            split_point = total_tokens / 2;
+            first_half_size = split_point;
+            second_half_size = total_tokens - split_point;
+        }
+
+        const float* visual_data = visual_features.data<const float>();
+        ov::Tensor visual_first_half;
+        ov::Tensor visual_second_half;
+
+        if (use_splitting) {
+            visual_first_half = ov::Tensor(ov::element::f32,
+                                           {batch_size, first_half_size, feature_dim},
+                                           const_cast<float*>(visual_data));
+            visual_second_half =
```

Comment:
could you please align formatting with `visual_first_half`?

Reply:
updated.